src/feature/hs/hs_service.c

```diff
@@ -1301,8 +1303,87 @@ service_descriptor_new(void)
   return sdesc;
 }
 
-/* Move descriptor(s) from the src service to the dst service. We do this
- * during SIGHUP when we re-create our hidden services. */
+/* If two authorized clients are equal, return 1. Otherwise, return 0. */
+static int
+service_authorized_client_equal(const hs_service_authorized_client_t *client1,
+                                const hs_service_authorized_client_t *client2)
+{
+  tor_assert(client1);
+  tor_assert(client2);
+
+  /* If the client public keys are not equal, the whole structs are
+   * not equal. */
+  if (!tor_memeq(&client1->client_pk,
+                 &client2->client_pk,
+                 sizeof(curve25519_public_key_t))) {
+    return 0;
+  }
+  return 1;
+}
+
+/* If the list of hs_service_authorized_client_t's is different between
+ * src and dst, return 1. Otherwise, return 0. */
+static int
+service_authorized_client_config_equal(const hs_service_config_t *config1,
+                                       const hs_service_config_t *config2)
+{
+  int ret = 0;
+  int *idx_matched = NULL;
+
+  tor_assert(config1);
+  tor_assert(config2);
+  tor_assert(config1->clients);
+  tor_assert(config2->clients);
+
+  /* If the number of clients is different, it is obvious that the list
+   * changes. */
+  if (smartlist_len(config1->clients) != smartlist_len(config2->clients)) {
+    goto done;
+  }
+
+  /* An array that has the same length as the list of clients in config2.
+   * If the client in config2 is matched with some client in config1, the
+   * value of this array at the corresponding index will be 1. Otherwise,
+   * it will be 0. */
+  idx_matched =
+    tor_malloc_zero(sizeof(*idx_matched) * smartlist_len(config2->clients));
+
+  SMARTLIST_FOREACH_BEGIN(config1->clients,
```

Comment:
I don't like doing this quadratic loop here -- do we really need to do this?

Reply:
One option would be to sort the lists, then compare them in lockstep.

Come to think of it, we should probably also  scramble the lists before we actually use them, so a client doesn't learn anything from its position in the list.

Reply:
This is unfortunately needed... It will happen rarely that is when the service is HUP. 

The reason we need this is to detect if there are new client authorization configured for the service and if so, we need to build a new descriptor for the service and republish.

So for that, we have to go through all clients from both services object (old and new) and figure out if any of them changed (removed or added).

> Come to think of it, we should probably also scramble the lists before we actually use them, so a client doesn't learn anything from its position in the list.

Oh hmmmm, yeah I think that is not a stupid idea. We should scramble that list after loading the key file. I propose we do this after, noting down.

Reply:
I agree to scramble the list. It's safer to not let the client know the position.
Yes it's better to improve the running time. Do we need to do it in this branch?

Reply:
A separate branch is fine, but IMO it should happen in 0.3.5 (especially since it is easy to do)